File: Sources/FlowLokal/StatsStore.swift
import Foundation
import Combine

/// Kumulative Nutzungs-Statistiken (bleiben erhalten, auch wenn der Verlauf
/// gekappt wird). JSON in ~/Library/Application Support/shout/stats.json.
@MainActor
final class StatsStore: ObservableObject {

    struct Data: Codable {
        var totalWords = 0
        var totalDictations = 0
        var totalSeconds = 0.0
        var activeDays: [String] = []   // "yyyy-MM-dd"
    }

    @Published private(set) var data = Data()

    private let fileURL: URL

    init() {
        fileURL = StoreIO.directory().appendingPathComponent("stats.json")
        if let decoded = StoreIO.load(Data.self, from: fileURL) { data = decoded }
    }

    func record(words: Int, seconds: Double, date: Date = Date()) {
        guard words > 0 else { return }
        data.totalWords += words
        data.totalDictations += 1
        data.totalSeconds += max(0, seconds)
        let key = Self.dayKey(date)
        if !data.activeDays.contains(key) { data.activeDays.append(key) }
        save()
    }

    var averageWPM: Int {
        guard data.totalSeconds > 1 else { return 0 }
        return Int((Double(data.totalWords) / (data.totalSeconds / 60)).rounded())
    }

    /// Aktueller Streak in Tagen (bis heute oder gestern zurück).
    var currentStreak: Int {
        let days = Set(data.activeDays)
        let cal = Calendar.current
        var streak = 0
        var day = cal.startOfDay(for: Date())
        // Wenn heute noch nichts, ab gestern zählen.
        if !days.contains(Self.dayKey(day)) { day = cal.date(byAdding: .day, value: -1, to: day)! }
        while days.contains(Self.dayKey(day)) {
            streak += 1
            day = cal.date(byAdding: .day, value: -1, to: day)!
        }
        return streak
    }

    var longestStreak: Int {
        let keys = data.activeDays.sorted()
        guard !keys.isEmpty else { return 0 }
        let cal = Calendar.current
        let dates = keys.compactMap { Self.dayFormatter.date(from: $0) }
        var longest = 1, run = 1
        for i in 1..<max(1, dates.count) {
            if let prev = cal.date(byAdding: .day, value: 1, to: dates[i - 1]),
               cal.isDate(prev, inSameDayAs: dates[i]) {
                run += 1
            } else {
                run = 1
            }
            longest = max(longest, run)
        }
        return longest
    }

    func isActive(_ date: Date) -> Bool {
        Set(data.activeDays).contains(Self.dayKey(date))
    }

    /// Ersetzt die Statistik-Daten (für Import).
    func replaceData(_ newData: Data) {
        data = newData
        save()
    }

    // MARK: - Persistenz

    private func save() {
        StoreIO.save(data, to: fileURL)
    }

    // MARK: - Helfer

    static func dayKey(_ date: Date) -> String { dayFormatter.string(from: date) }
    private static let dayFormatter: DateFormatter = {
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd"; f.locale = Locale(identifier: "en_US_POSIX"); return f
    }()
}

File: windows/src/Shout/Core/StatsStore.cs
using System.Text.Json.Serialization;

namespace Shout.Core;

/// <summary>
/// Kumulative Nutzungs-Statistiken (bleiben erhalten, auch wenn der Verlauf
/// gekappt wird) — stats.json, Feldnamen kompatibel zum Mac.
/// </summary>
public sealed class StatsStore
{
    public sealed class StatsData
    {
        [JsonPropertyName("totalWords")] public int TotalWords { get; set; }
        [JsonPropertyName("totalDictations")] public int TotalDictations { get; set; }
        [JsonPropertyName("totalSeconds")] public double TotalSeconds { get; set; }
        [JsonPropertyName("activeDays")] public List<string> ActiveDays { get; set; } = new();   // "yyyy-MM-dd"
    }

    public StatsData Data { get; private set; } = new();

    public StatsStore()
    {
        Data = StoreIO.Load<StatsData>("stats.json") ?? new StatsData();
    }

    public void Record(int words, double seconds)
    {
        if (words <= 0) return;
        Data.TotalWords += words;
        Data.TotalDictations += 1;
        Data.TotalSeconds += Math.Max(0, seconds);
        var key = DayKey(DateTime.Now);
        if (!Data.ActiveDays.Contains(key)) Data.ActiveDays.Add(key);
        Save();
    }

    public int AverageWpm =>
        Data.TotalSeconds > 1
            ? (int)Math.Round(Data.TotalWords / (Data.TotalSeconds / 60))
            : 0;

    /// <summary>Aktueller Streak in Tagen (bis heute oder gestern zurück).</summary>
    public int CurrentStreak
    {
        get
        {
            var days = new HashSet<string>(Data.ActiveDays);
            var day = DateTime.Today;
            if (!days.Contains(DayKey(day))) day = day.AddDays(-1);
            var streak = 0;
            while (days.Contains(DayKey(day)))
            {
                streak++;
                day = day.AddDays(-1);
            }
            return streak;
        }
    }

    /// <summary>Längster jemals erreichter Streak (für die Statistik-Seite).</summary>
    public int LongestStreak
    {
        get
        {
            var days = Data.ActiveDays
                .Select(d => DateTime.TryParse(d, out var parsed) ? parsed.Date : (DateTime?)null)
                .Where(d => d.HasValue)
                .Select(d => d!.Value)
                .Distinct()
                .OrderBy(d => d)
                .ToList();

            var best = 0;
            var run = 0;
            DateTime? previous = null;
            foreach (var day in days)
            {
                run = previous.HasValue && (day - previous.Value).Days == 1 ? run + 1 : 1;
                best = Math.Max(best, run);
                previous = day;
            }
            return best;
        }
    }

    /// <summary>War an diesem Tag mindestens ein Diktat? (Aktivitäts-Kalender)</summary>
    public bool IsActive(DateTime day) => Data.ActiveDays.Contains(DayKey(day));

    /// <summary>Ersetzt die Statistik-Daten (für Import).</summary>
    public void ReplaceData(StatsData newData)
    {
        Data = newData;
        Save();
    }

    private void Save() => StoreIO.Save(Data, "stats.json");

    public static string DayKey(DateTime date) => date.ToString("yyyy-MM-dd");
}
